src/platform/Linux/MdnsImpl.cpp

```diff
@@ -0,0 +1,665 @@
+        ChipLogProgress(DeviceLayer, "Avahi client registered");
+        mClient = client;
+        mGroup  = avahi_entry_group_new(client, HandleGroupState, this);
+        if (mGroup == nullptr)
+        {
+            ChipLogError(DeviceLayer, "Failed to create avahi group: %s", avahi_strerror(avahi_client_errno(client)));
+            mInitCallback(mAsyncReturnContext, CHIP_ERROR_OPEN_FAILED);
+        }
+        else
+        {
+            mInitCallback(mAsyncReturnContext, CHIP_NO_ERROR);
+        }
+        break;
+    case AVAHI_CLIENT_FAILURE:
+        ChipLogError(DeviceLayer, "Avahi client failure");
+        mErrorCallback(mAsyncReturnContext, CHIP_ERROR_INTERNAL);
+        break;
+    case AVAHI_CLIENT_S_COLLISION:
+    case AVAHI_CLIENT_S_REGISTERING:
+        ChipLogProgress(DeviceLayer, "Avahi re-register required");
+        mErrorCallback(mAsyncReturnContext, CHIP_ERROR_MDNS_COLLISSION);
+        if (mGroup != nullptr)
+        {
+            avahi_entry_group_reset(mGroup);
+        }
+        break;
+    case AVAHI_CLIENT_CONNECTING:
+        ChipLogProgress(DeviceLayer, "Avahi connecting");
+        break;
+    }
+}
+
+void MdnsAvahi::HandleGroupState(AvahiEntryGroup * group, AvahiEntryGroupState state, void * context)
+{
+    static_cast<MdnsAvahi *>(context)->HandleGroupState(group, state);
+}
+
+void MdnsAvahi::HandleGroupState(AvahiEntryGroup * group, AvahiEntryGroupState state)
+{
+    switch (state)
+    {
+    case AVAHI_ENTRY_GROUP_ESTABLISHED:
+        ChipLogProgress(DeviceLayer, "Avahi group established");
+        break;
+    case AVAHI_ENTRY_GROUP_COLLISION:
+        ChipLogError(DeviceLayer, "Avahi group collission");
+        mErrorCallback(mAsyncReturnContext, CHIP_ERROR_MDNS_COLLISSION);
+        break;
+    case AVAHI_ENTRY_GROUP_FAILURE:
+        ChipLogError(DeviceLayer, "Avahi group internal failure %s",
+                     avahi_strerror(avahi_client_errno(avahi_entry_group_get_client(mGroup))));
+        mErrorCallback(mAsyncReturnContext, CHIP_ERROR_INTERNAL);
+        break;
+    case AVAHI_ENTRY_GROUP_UNCOMMITED:
+    case AVAHI_ENTRY_GROUP_REGISTERING:
+        break;
+    }
+}
+
+static const char * GetProtocolString(MdnsServiceProtocol protocol)
+{
+    return protocol == MdnsServiceProtocol::kMdnsProtocolUdp ? "_udp" : "_tcp";
+}
+
+static std::string GetFullType(const char * type, MdnsServiceProtocol protocol)
+{
+    std::ostringstream typeBuilder;
+    typeBuilder << type << "." << GetProtocolString(protocol);
+    return typeBuilder.str();
+}
+
+static AvahiStringList * MakeAvahiStringListFromTextEntries(TextEntry * entries, size_t size)
+{
+    AvahiStringList * strList = avahi_string_list_new(nullptr, nullptr);
+
+    for (size_t i = 0; i < size; i++)
+    {
+        uint8_t buf[kMdnsTypeMaxSize];
+        size_t keySize   = strnlen(entries[i].mKey, kMdnsTypeMaxSize);
+        size_t valueSize = 0;
+
+        memcpy(buf, entries[i].mKey, keySize);
+        if (size < sizeof(buf))
```

Comment:
size -> keySize?

I don't understand this condition, can you write a comment? Should it be an error if we can't fit the value?

Is this just `snprintf(buf, sizeof(buf), "%s=%s", ...)` ?

Reply:
mDNS allows using arbitrary binary data rather than plain text so `\0` is a valid character in the data entry.